rfq/stream.go

```diff
@@ -0,0 +1,306 @@
+
+		ContextGuard: &fn.ContextGuard{
+			DefaultTimeout: DefaultTimeout,
+			Quit:           make(chan struct{}),
+		},
+	}, nil
+}
+
+// handleIncomingQuoteRequest handles an incoming quote request peer message.
+func (h *StreamHandler) handleIncomingQuoteRequest(
+	wireMsg rfqmsg.WireMessage) error {
+
+	quoteRequest, err := rfqmsg.NewRequestMsgFromWire(wireMsg)
+	if err != nil {
+		return fmt.Errorf("unable to create a quote request message "+
+			"from a wire message: %w", err)
+	}
+
+	log.Debugf("Stream handling incoming message (msg_type=%T, "+
+		"msg_id=%s, origin_peer=%s)", quoteRequest,
+		quoteRequest.ID.String(), quoteRequest.MsgPeer())
+
+	// Send the quote request to the RFQ manager.
+	var msg rfqmsg.IncomingMsg = quoteRequest
+	sendSuccess := fn.SendOrQuit(h.cfg.IncomingMessages, msg, h.Quit)
+	if !sendSuccess {
+		return fmt.Errorf("RFQ stream handler shutting down")
+	}
+
+	return nil
+}
+
+// handleIncomingQuoteAccept handles an incoming quote accept peer message.
+func (h *StreamHandler) handleIncomingQuoteAccept(
+	wireMsg rfqmsg.WireMessage) error {
+
+	quoteAccept, err := rfqmsg.NewAcceptFromWireMsg(wireMsg)
+	if err != nil {
+		return fmt.Errorf("unable to create a quote accept message "+
+			"from a wire message: %w", err)
+	}
+
+	log.Debugf("Stream handling incoming message (msg_type=%T, "+
+		"origin_peer=%s)", quoteAccept, quoteAccept.MsgPeer())
+
+	// TODO(ffranr): At this point, we need to validate that the quote
+	//  accept message is valid. We need to check that the quote accept
+	//  message is associated with a valid quote request.
+
+	// Send the message to the RFQ manager.
+	var msg rfqmsg.IncomingMsg = quoteAccept
+	sendSuccess := fn.SendOrQuit(h.cfg.IncomingMessages, msg, h.Quit)
+	if !sendSuccess {
+		return fmt.Errorf("RFQ stream handler shutting down")
+	}
+
+	return nil
+}
+
+// handleIncomingQuoteReject handles an incoming quote reject peer message.
+func (h *StreamHandler) handleIncomingQuoteReject(
+	wireMsg rfqmsg.WireMessage) error {
+
+	quoteReject, err := rfqmsg.NewQuoteRejectFromWireMsg(wireMsg)
+	if err != nil {
+		return fmt.Errorf("unable to create a quote reject message "+
+			"from a wire message: %w", err)
+	}
+
+	log.Debugf("Stream handling incoming message (msg_type=%T, "+
+		"origin_peer=%s)", quoteReject, quoteReject.MsgPeer())
+
+	// TODO(ffranr): At this point, we need to validate that the quote
+	//  reject message is valid. We need to check that the quote reject
+	//  message is associated with a valid quote request.
+
+	// Send the message to the RFQ manager.
+	var msg rfqmsg.IncomingMsg = quoteReject
+	sendSuccess := fn.SendOrQuit(h.cfg.IncomingMessages, msg, h.Quit)
+	if !sendSuccess {
+		return fmt.Errorf("RFQ stream handler shutting down")
+	}
+
+	return nil
+}
+
+// handleIncomingWireMessage handles an incoming wire message.
+func (h *StreamHandler) handleIncomingWireMessage(
+	wireMsg rfqmsg.WireMessage) error {
+
+	switch wireMsg.MsgType {
+	case rfqmsg.MsgTypeRequest:
```

Comment:
nit: There's a fair amount of duplicate code here for each message type. Could be simplified to something like:

```go
// parse incoming message
var msg rfqmsg.IncomingMsg

switch wireMsg.MsgType {
  case rfqmsg.MsgTypeRequest:
    quoteReject, err := rfqmsg.NewQuoteRejectFromWireMsg(wireMsg)
    // handle err
    // debug log message
    
    msg = rfqmsg.IncomingMsg(quoteReject)
    
    case rfqmsg.MsgTypeAccept:
    // ....
    
    // handle final case + default

}

sendSuccess := fn.SendOrQuit(h.cfg.IncomingMessages, msg, h.Quit)
if !sendSuccess {
	return fmt.Errorf("RFQ stream handler shutting down")
}
```

I think the only missing piece there is a `String()` method for `rfq.MsgType` to help with accurate log messages.

Reply:
I've simplified this code in the latest commits. I wasn't sure which way the `stream` service was going to go, so I made it verbose and repetitive. But now, I think that `stream` will only handle send/recv and message type casting.